chalk-solve/src/clauses/builtin_traits/unsize.rs

```diff
@@ -0,0 +1,449 @@
+    interner: &I,
+    v: &Binders<impl Visit<I> + HasInterner>,
+    parameters: &HashSet<usize>,
+) -> bool {
+    let mut visitor = ParameterOccurenceCheck {
+        interner,
+        parameters,
+    };
+    v.visit_with(&mut visitor, DebruijnIndex::INNERMOST) == FindAny::FOUND
+}
+
+fn principal_id<'a, I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    bounds: &'a Binders<QuantifiedWhereClauses<I>>,
+) -> Option<TraitId<I>> {
+    let interner = db.interner();
+
+    let principal_id = bounds
+        .skip_binders()
+        .iter(interner)
+        .next()
+        .expect("Expected trait object to have at least one trait bound")
+        .trait_id()?;
+
+    if db.trait_datum(principal_id).is_auto_trait() {
+        None
+    } else {
+        Some(principal_id)
+    }
+}
+
+fn auto_trait_ids<'a, I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    bounds: &'a Binders<QuantifiedWhereClauses<I>>,
+) -> impl Iterator<Item = TraitId<I>> + 'a {
+    let interner = db.interner();
+    // all trait ref where clauses after the principal are auto traits
+    let to_skip = if principal_id(db, bounds).is_some() {
+        1
+    } else {
+        0
+    };
+    bounds
+        .skip_binders()
+        .iter(interner)
+        .skip(to_skip)
+        .filter_map(|clause| clause.trait_id())
+}
+
+pub fn add_unsize_program_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    trait_ref: &TraitRef<I>,
+    _ty: &TyData<I>,
+) {
+    let interner = db.interner();
+
+    let source_ty = trait_ref.self_type_parameter(interner);
+    let target_ty = trait_ref
+        .substitution
+        .at(interner, 1)
+        .assert_ty_ref(interner);
+
+    let unsize_trait_id = trait_ref.trait_id;
+
+    // N.B. here rustc asserts that `TraitRef` is not a higher-ranked bound
+    // i.e. `for<'a> &'a T: Unsize<dyn Trait+'a>` is never provable.
+    //
+    // In chalk it would be awkward to implement and I am not sure
+    // there is a need for it, the original comment states that this restriction
+    // could be lifted.
+    //
+    // for more info visit `fn assemble_candidates_for_unsizing` and
+    // `fn confirm_builtin_unisize_candidate` in rustc.
+
+    match (source_ty.data(interner), target_ty.data(interner)) {
+        // dyn Trait + AutoX + 'a -> dyn Trait + AutoY + 'b
+        (
+            TyData::Dyn(DynTy {
+                bounds: bounds_a,
+                lifetime: lifetime_a,
+            }),
+            TyData::Dyn(DynTy {
+                bounds: bounds_b,
+                lifetime: lifetime_b,
+            }),
+        ) => {
+            let principal_a = principal_id(db, bounds_a);
+            let principal_b = principal_id(db, bounds_b);
+
+            let auto_trait_ids_a: Vec<_> = auto_trait_ids(db, bounds_a).collect();
+            let auto_trait_ids_b: Vec<_> = auto_trait_ids(db, bounds_b).collect();
+
+            let may_apply = principal_a == principal_b
```

Comment:
So, really, this is just checking if the trait bounds of `a` are a subset of the trait bounds on `b`. So we really don't need to split out auto traits and the principal trait here, right?

Reply:
Somewhat separately: I know this is mostly copied from the rustc code, and as far as I can tell, it matches. But this feels backwards to me. If the comment represents the truth that we're trying to go from `dyn Trait + AutoX + 'a -> dyn Trait + AutoY + 'b` (and `source_ty` is `dyn Trait AutoX + 'a`), then shouldn't the bounds on `b` need to be a subset of `a`?

Reply:
We do need to split them, because 
```
dyn Principal + Auto -> dyn Principal  
```
if a valid coercion, but
```
dyn Trait + Send -> dyn Send
```
is not. I don't know a reason for that, but that's how it is ¯\_(ツ)_/¯.

As for the second comment, that is what's happening: for all auto bounds of b we find a matching auto bound of a.